dask/dataframe/partitionquantiles.py

```diff
@@ -429,6 +442,11 @@ def percentiles_summary(df, num_old, num_new, upsample, state):
     except (TypeError, NotImplementedError):
         vals, _ = _percentile(array_safe(data, data.dtype), qs, interpolation)
 
+    # FIXME: `ExtensionArrays` don't support `tolist` in pandas<1.4 so we must
+    # cast to a ndarray.
+    if not PANDAS_GT_140 and is_extension_array_dtype(vals):
+        vals = vals.astype("object")
+
```

Comment:
Hmm, this seems like a shame. I'm not actually sure why we need to convert to a list in the first place. Do you know if they could be dropped somehow?

Reply:
Tried dropping the `tolist()` calls and seems like the shuffle tests are un-impacted locally, will push this change up and see if anything crops up in CI

Reply:
The failures that crop up from this change are on GPU, and are related to the fact that we later try to construct a NumPy array from `vals`, which fails when `vals` is a CuPy array:

```
dask/dataframe/partitionquantiles.py:321: in process_val_weights
    vals = np.array(vals)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
```

I propose that we make it more explicit that a host sync is required here to get things working, that we can ideally try to remove in a follow-up. In general, I think it would be worth it to comb through the GPU-specific code in this module, as it looks like this block

https://github.com/dask/dask/blob/f09b0897a49cfd98c854feb06c5b179ab3c6fbdb/dask/dataframe/partitionquantiles.py#L432-L440

Is no longer being accessed since `data` is seemingly always going to be a frame (cc @pentschev).

Reply:
If `vals` is just a CuPy array and there are no further operations that aren't yet support by CuPy, the solution may be as simple as changing that line to:

```python
from dask.array.utils import array_safe
vals = array_safe(vals, like=vals)
```

Could you try that @charlesbluca ?

Reply:
That does unblock that failure, thanks @pentschev 🙂 unfortunately, it looks like shortly after successfully creating that array, we end up using it to create a Pandas object, which ends up raising the same error:

https://github.com/dask/dask/blob/b4e20d45bf2473df16995d9c689c7f77755add17/dask/dataframe/partitionquantiles.py#L379-L386

Reply:
Maybe the `meta_lib_from_array` dispatcher Rick added in https://github.com/dask/dask/pull/9579 would be useful there?